### PR TITLE
Fixing the sourcelink failure for embeded files

### LINKED ISSUE #33097: For .net 5 preview1, the CoreFX symbols are missing for IL symbols.
Affected Build: Master/29827.154 (16.6.0 preview 1.0) + CLI/SDK:5.0.100-preview.1.20125.9

**Steps to reproduce**
Testing on Windows with VS -

1. Goto Tools -> options  -> Debugging
Under General
	a. Uncheck 'Enable Just My Code'
	b. Check 'Suppress JIT optimization on module load (Managed only)'
Under Symbols
	a. Check 'Microsoft Symbol Servers'
2. Create a new ASP.NET Razor project by doing
mkdir razor
cd razor
dotnet new razor

3. Double click on the ‘Startup.cs’ file in the ‘Explorer’ pane to bring up that source code. 
4.  Add the following line to the Startup.cs file at the beginning of the ‘Configure’ method (at the bottom of the file). When using VS, you can add this line of code on Program.cs file
 
	Console.WriteLine("x " + 3);
 
5. Place a breakpoint on that line (F9). 
6. Run the application until it hit that breakpoint (F5).   It will load symbols from the Microsoft symbol server, and may take a few seconds to hit the breakpoint.  You should see 20-30 lines of ‘Loaded XXX Symbols loaded’.

7. Step into the concatination method (F11).  Select 'Download Source and Continue Debugging' from the popup
This tests CoreCLR symbols   AT THIS POINT IT SHOULD SHOW YOU THE SOURCE CODE FOR Concat(). 

8. Step out of Concat (Shift F11). Then step into the WriteLine method(F11). You may see the pop up as above, select the first option from it. 
This tests CoreFX symbols  AT THIS POINT IT SHOULD SHOW YOU THE SOURCE CODE FOR Console.WriteLine(). 

**Expected result:**

CoreFX symbols are displayed in step8.

**Actual result:**

CoreFX symbols are missing in step8.
image

Dotnet –info:

.NET Core SDK (reflecting any global.json):
 Version:   5.0.100-preview.1.20125.9
 Commit:    12e88ed2e5

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18362
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.100-preview.1.20125.9\

Host (useful for support):
  Version: 5.0.0-preview.1.20120.5
  Commit:  3c523a6a7a

.NET Core SDKs installed:
  3.1.200-preview-014977 [C:\Program Files\dotnet\sdk]
  5.0.100-preview.1.20125.9 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.0-preview.1.20124.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.0-preview.1.20120.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.0-preview.1.20122.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download

### PR DESCRIPTION
**Pull request by Anipik**
Generating files before sourcelink calculates the embedded files
Fixes https://github.com/dotnet/runtime/issues/33097

### CODE REVIEW

**Comment by danmosemsft** [comment]
Please port to preview 2 branch also. You can consider it approved if risk is acceptable to you and @ericstj

**Comment by safern** [comment]
cc: @dotnet/runtime-infrastructure

File: Directory.Build.props [hunk 1]
```diff
@@ -54,7 +54,13 @@
     <ManPagesDir>$([MSBuild]::NormalizeDirectory('$(DocsDir)', 'manpages'))</ManPagesDir>
     <CoreLibSharedDir>$([MSBuild]::NormalizeDirectory('$(LibrariesProjectRoot)', 'System.Private.CoreLib', 'src'))</CoreLibSharedDir>
   </PropertyGroup>
-
+  
+  <!-- SourceLink -->
+  <PropertyGroup>
+    <!-- This is needed to embed build-time generated sources such as eventing and resource files to sourcelink PDBs. -->
+    <EmbedUntrackedSources>true</EmbedUntrackedSources>
```
Comment (safern) [inline Directory.Build.props]: Should this be in `Versions.props`? We usually enable/disable features from arcade there. @tmat ?

Comment (Anipik) [reply Directory.Build.props]: Sourcelink lives in a separate repo. https://github.com/dotnet/sourcelink. We can still move to ```versions.props``` if we want.

**Review by safern — COMMENTED** [review]
Could you also run an official build with this changes and enabling the source link validation? 

I think we should enable the source link validation as part of this change to catch these errors earlier.

File: Directory.Build.props (hunk 1 above)
Comment (safern) [reply Directory.Build.props]: Oh, cool. Then this is fine here.

**Comment by Anipik** [comment]
> Could you also run an official build with this changes and enabling the source link validation?

yeah i will do that.

**Comment by Anipik** [comment]
https://dev.azure.com/dnceng/internal/_build/results?buildId=557428

official build

File: Directory.Build.props (hunk 1 above)
Comment (tmat) [reply Directory.Build.props]: It shouldn't be necessary here. It's already set by Arcade here: [https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Arcade.Sdk/tools/ProjectDefaults.props#L18](https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Arcade.Sdk/tools/ProjectDefaults.props#L18). Is this props file not imported for some reason?

---
In reply to: [391824522](https://github.com/dotnet/runtime/pull/33521#discussion_r391824522) [](ancestors = 391824522)

Comment (safern) [reply Directory.Build.props]: It should be imported.

File: Directory.Build.targets [hunk 2]
```diff
@@ -20,4 +20,7 @@
     <NETCoreAppMaximumVersion>$(MajorVersion).$(MinorVersion)</NETCoreAppMaximumVersion>
   </PropertyGroup>
 
+  <Target Name="GeneratingFilesBeforeRunningSourceLink"
+          BeforeTargets="SetEmbeddedFilesFromSourceControlManagerUntrackedFiles"
+          DependsOnTargets="GenerateAssemblyInfo;GenerateInternalsVisibleToFile" />
```
Comment (tmat) [inline Directory.Build.targets]: This shouldn't be necessary as long as all targets that add a `Compile` item have `BeforeTargets="BeforeCompile;CoreCompile"`.

Comment (hoyosjs) [reply Directory.Build.targets]: GenerateAssemblyInfo in the SDK had the issue though. You fixed it in dotnet/sdk#10613. Until we pick that SDK, a better workaround would be what was suggested in dotnet/sourcelink#572:
```xml
<Project>
  <PropertyGroup>
    <TargetFrameworkMonikerAssemblyAttributesPath>$([System.IO.Path]::Combine('$(IntermediateOutputPath)','$(TargetFrameworkMoniker).AssemblyAttributes$(DefaultLanguageSourceExtension)'))</TargetFrameworkMonikerAssemblyAttributesPath>
  </PropertyGroup>
  <ItemGroup>
    <EmbeddedFiles Include="$(GeneratedAssemblyInfoFile)"/>
  </ItemGroup>
</Project>
```

**Review by hoyosjs — APPROVED** [review]
That one nit. Also, as long as Mono's SPC build also uses arcade defaults sourcelink should work.

File: eng/common/templates/post-build/post-build.yml [hunk 3]
```diff
@@ -1,5 +1,5 @@
 parameters:
-  enableSourceLinkValidation: false
```
[the remark this answers is not in the document]
Comment (safern) [reply eng/common/templates/post-build/post-build.yml]: Yes, we need to revert this change.

**Comment by Anipik** [comment]
sourcelink validation succeeded https://dev.azure.com/dnceng/internal/_build/results?buildId=558701

File: Directory.Build.targets [hunk 4]
```diff
@@ -20,4 +20,11 @@
     <NETCoreAppMaximumVersion>$(MajorVersion).$(MinorVersion)</NETCoreAppMaximumVersion>
   </PropertyGroup>
 
+  <PropertyGroup>
```
Comment (tmat) [inline Directory.Build.targets]: ><PropertyGroup> [](start = 2, length = 15)

It'd be good to add a comment that this is a workaround for a bug in the SDK and link to the issue that fixes it.

**Review by tmat — APPROVED** [review]
:shipit:

File: Directory.Build.targets (hunk 4 above)
Comment (safern) [reply Directory.Build.targets]: But the workaround is adding the file to EmbeddedFiles right?

Comment (hoyosjs) [reply Directory.Build.targets]: Yes, any AssemblyAttributes files that got generated out of order before this flows through https://github.com/dotnet/sdk/pull/10613

**Comment by hoyosjs** [comment]
> Please port to preview 2 branch also. You can consider it approved if risk is acceptable to you and @ericstj Eric St. John FTE

The prev2 branch wouldn't pass with this change as-is as the arcade version on the branch doesn't contain dotnet/arcade@527179abfbb5cd738035c4ac2d2566c571ce85db

**Comment by Anipik** [comment]
I talked with @ericstj offline. we dont need to port this preview2 as the embedded files dont contain any actual  code, these are just typeforwards or attributes. we can wait till preview 3

File: Directory.Build.targets [hunk 5]
```diff
@@ -20,4 +20,15 @@
...
+  <!--
+    All source inputs to the compiler should be generated before BeforeCompile target. Sdk is not
```
Comment (safern) [inline Directory.Build.targets]: Shouldn't this property always be needed? It seems like the comment needs to be on the ItemGroup that is appending to `EmbeddedFiles`

Comment (Anipik) [reply Directory.Build.targets]: No, we are just computing the path of generated file here and then including it that path in embedded files. Bu ti think the comment i think still should be moved to itemgroup

Comment (safern) [reply Directory.Build.targets]: I meant that even when the issue is fixed, we will still need to set `GeneratedAssemblyInfoFile`, correct?

**Review by safern — APPROVED** [review]

File: Directory.Build.targets (hunk 5 above)
Comment (hoyosjs) [reply Directory.Build.targets]: @safern why do we need the property?

Comment (safern) [reply Directory.Build.targets]: You're right. I got confused because of libraries that we write more stuff to that file and define our own property.

**Comment by Anipik** [comment]
@ericstj @safern @ViktorHofer any insight about the failure.
I was not able to find anything good in the log https://helix.dot.net/api/2019-06-17/jobs/3b8f6fb9-6ab6-4ba6-bec5-c9f313c1dc0f/workitems/baseservices.threading/console

This doesnt seem to be related to this pr but i didnt have much info about the error to create an issue for this.

**Comment by danmosemsft** [comment]
@Anipik you might have to try to repro locally? 😕

**Comment by safern** [comment]
I fixed this here: https://github.com/dotnet/runtime/pull/33580 you just have to rerun the whole pipeline, or you can ignore and merge.

**Pull request merged**